## MCE to Determine Highest-Risk Intersection most in Need of Leading Pedestrian Interval Signals
![NYC Streetlight Intersection](images/street_NYC.png)
<details>
  <summary>Introduction!</summary>
    
###### With the fast-paced hustle-and-bustle movement of pedestrians, vehicles and bicycles that all need to get from point A to point B in a ‘New York Minute’, city planners and lawmakers alike know the dangers traffic accidents present in the city that never sleeps. In 2014, New York City passed a new policy known as Vision Zero, which seeks to eliminate as many traffic accidents and pedestrian casualties as possible. In 2014 alone, 3,000 New Yorkers were seriously injured, and more than 200 were killed in traffic crashes. The leading cause of injury-related deaths for children under 14 and the second leading cause for senior citizens is being struck by moving vehicles. New York City has taken many proactive measures in limiting traffic collisions and pedestrian deaths by implementing Leading Pedestrian Interval Signals(LPIS) throughout the five boroughs. These signals allow pedestrians a head start to walk across the crosswalk before the green is given to vehicles, giving drivers more visibility of possible pedestrians in the process. There are currently over 4,000 of LPISs installed and working, but the goal of this research is to use a Multi-Criteria Evaluation(MCE) of New York City’s streets and intersections to determine the intersections most in need of LPIS installation.



NOTE: This notebook and subsequent notebooks will serve as a guide on how to clean and refine large datasets, conduct spatial analysis in PostGIS as well as python, standardize variables of interest, and present final data findings in interactive and aesthetic maps.

contact:Steven.Nacinovich27@myhunter.cuny.edu 

## Part 1: Cleaning our Collisions and pedestrian data-sets

 Data retrieved from:
 https://data.cityofnewyork.us/Public-Safety/Motor-Vehicle-Collisions-Crashes/h9gi-nx95 
 &
 https://data.cityofnewyork.us/Public-Safety/Motor-Vehicle-Collisions-Person/f55k-p6yu
    
Collision crash table contains information from all police reported motor vehicle collisions since 2014, when Vision Zero was first created. The police report is required to be filled out when someone is injured, killed, or there is more than $1,000 dollars in reported damages. There are 1.68 million records in total.

The person crash table contains information for each person involved in motor vehicle collisions since 2016. There are 3.92 million records in total.


In [2]:
%matplotlib inline

# Import libraries used below
import pandas as pd
import geopandas as gpd
import numpy as np
import json
from sqlalchemy import create_engine
from shapely import wkt

Reading in data using pandas library

In [ ]:
collision = pd.read_csv("F:\GeoCompII\project\Data\City_OpenData\Motor_Vehicle_Collisions_-_Crashes.csv")
pedestrian = pd.read_csv("F:\GeoCompII\project\Data\City_OpenData\Motor_Vehicle_Collisions_-_Person.csv")

In [4]:
# view the collision data
collision.head()

CRASH DATE CRASH TIME    BOROUGH ZIP CODE  LATITUDE  LONGITUDE  \
0  03/03/2020       0:00   BROOKLYN    11208  40.68992  -73.87261   
1  03/13/2019      10:15  MANHATTAN    10004       NaN        NaN   
2  04/02/2015       1:15        NaN      NaN       NaN        NaN   
3  04/02/2015      22:30        NaN      NaN       NaN        NaN   
4  03/03/2020      11:20  MANHATTAN    10036       NaN        NaN   

                     LOCATION                    ON STREET NAME  \
0  POINT (-73.87261 40.68992)  HEMLOCK STREET                     
1                         NaN                               NaN   
2                         NaN                               NaN   
3                         NaN  BRUCKNER BOULEVARD                 
4                         NaN  W 46th STREET                      

                  CROSS STREET NAME                           OFF STREET NAME  \
0                    JAMAICA AVENUE                                       NaN   
1                               NaN  10        South street                     
2                               NaN                                       NaN   
3  ZEREGA AVENUE                                                          NaN   
4                       12TH AVENUE                                       NaN   

   ...   CONTRIBUTING FACTOR VEHICLE 2  CONTRIBUTING FACTOR VEHICLE 3  \
0  ...                             NaN                            NaN   
1  ...                             NaN                            NaN   
2  ...  Driver Inattention/Distraction                            NaN   
3  ...                     Unspecified                            NaN   
4  ...                     Unspecified                            NaN   

   CONTRIBUTING FACTOR VEHICLE 4  CONTRIBUTING FACTOR VEHICLE 5  COLLISION_ID  \
0                            NaN                            NaN       4297861   
1                            NaN                            NaN       4096508   
2                            NaN                            NaN       3196332   
3                            NaN                            NaN       3197387   
4                            NaN                            NaN       4297078   

                   VEHICLE TYPE CODE 1  VEHICLE TYPE CODE 2  \
0                                  NaN                  NaN   
1  Station Wagon/Sport Utility Vehicle                  NaN   
2                    PASSENGER VEHICLE    PASSENGER VEHICLE   
3                              UNKNOWN    PASSENGER VEHICLE   
4                            Box Truck            Box Truck   

   VEHICLE TYPE CODE 3 VEHICLE TYPE CODE 4 VEHICLE TYPE CODE 5  
0                  NaN                 NaN                 NaN  
1                  NaN                 NaN                 NaN  
2                  NaN                 NaN                 NaN  
3                  NaN                 NaN                 NaN  
4                  NaN                 NaN                 NaN  

[5 rows x 29 columns]

In [5]:
# view pedestrian data
pedestrian.head()

UNIQUE_ID  COLLISION_ID  CRASH_DATE CRASH_TIME  \
0   10252788       4230640  10/25/2019      17:42   
1   10253512       4230733  10/10/2019      19:45   
2   10250662       4230235  10/26/2019      10:40   
3   10252894       4231115  10/26/2019      18:45   
4    6650180       3565527  11/21/2016      13:05   

                              PERSON_ID PERSON_TYPE PERSON_INJURY  VEHICLE_ID  \
0  0278267c-0b21-470f-8a54-da71b7361448    Occupant   Unspecified  19142923.0   
1  81161123-516c-438d-8f29-a107d2783d85    Occupant   Unspecified  19143291.0   
2  b5da9aa6-3230-4203-bbde-3b6f9248ac31    Occupant       Injured  19141868.0   
3  4ed210b6-678c-4fe3-ae8c-3ecd30eae7aa  Pedestrian       Injured         NaN   
4                               2782525    Occupant   Unspecified         NaN   

   PERSON_AGE     EJECTION  ...        BODILY_INJURY  \
0         0.0  Not Ejected  ...       Does Not Apply   
1         NaN          NaN  ...                  NaN   
2        11.0  Not Ejected  ...          Entire Body   
3        30.0          NaN  ...  Knee-Lower Leg Foot   
4         NaN          NaN  ...                  NaN   

                                 POSITION_IN_VEHICLE SAFETY_EQUIPMENT  \
0  Any person in the rear of a station wagon, pic...          Unknown   
1                                                NaN              NaN   
2  Middle rear seat, or passenger lying across a ...          Unknown   
3                                             Driver             None   
4                                                NaN              NaN   

                                        PED_LOCATION               PED_ACTION  \
0                                                NaN                      NaN   
1                                                NaN                      NaN   
2                                                NaN                      NaN   
3  Pedestrian/Bicyclist/Other Pedestrian at Inter...  Crossing Against Signal   
4                                                NaN                      NaN   

                            COMPLAINT         PED_ROLE  \
0                      Does Not Apply        Passenger   
1                                 NaN       Registrant   
2                            Whiplash        Passenger   
3  Fracture - Distorted - Dislocation       Pedestrian   
4                                 NaN  Notified Person   

                               CONTRIBUTING_FACTOR_1 CONTRIBUTING_FACTOR_2  \
0                                                NaN                   NaN   
1                                                NaN                   NaN   
2                                                NaN                   NaN   
3  Pedestrian/Bicyclist/Other Pedestrian Error/Co...               Texting   
4                                                NaN                   NaN   

  PERSON_SEX  
0          F  
1        NaN  
2          F  
3          M  
4        NaN  

[5 rows x 21 columns]

When Viewing the fields and data types in the objects Collision and Pedestrian, they each share one common field(COLLISION_ID). These files can be joined or merged to create one file that has all pedestrian injuries and vehicle collisions.
In this example, merge was used

In [6]:
df = collision.merge(pedestrian, how = 'left', left_on = 'COLLISION_ID', right_on = 'COLLISION_ID')
df.head()

CRASH DATE CRASH TIME    BOROUGH ZIP CODE  LATITUDE  LONGITUDE  \
0  03/03/2020       0:00   BROOKLYN    11208  40.68992  -73.87261   
1  03/13/2019      10:15  MANHATTAN    10004       NaN        NaN   
2  03/13/2019      10:15  MANHATTAN    10004       NaN        NaN   
3  03/13/2019      10:15  MANHATTAN    10004       NaN        NaN   
4  04/02/2015       1:15        NaN      NaN       NaN        NaN   

                     LOCATION                    ON STREET NAME  \
0  POINT (-73.87261 40.68992)  HEMLOCK STREET                     
1                         NaN                               NaN   
2                         NaN                               NaN   
3                         NaN                               NaN   
4                         NaN                               NaN   

  CROSS STREET NAME                           OFF STREET NAME  ...  \
0    JAMAICA AVENUE                                       NaN  ...   
1               NaN  10        South street                    ...   
2               NaN  10        South street                    ...   
3               NaN  10        South street                    ...   
4               NaN                                       NaN  ...   

         BODILY_INJURY                                POSITION_IN_VEHICLE  \
0  Knee-Lower Leg Foot                                                NaN   
1       Does Not Apply                                             Driver   
2       Does Not Apply  Middle front seat, or passenger lying across a...   
3                  NaN                                                NaN   
4                  NaN                                                NaN   

     SAFETY_EQUIPMENT                                       PED_LOCATION  \
0                 NaN  Pedestrian/Bicyclist/Other Pedestrian at Inter...   
1  Lap Belt & Harness                                                NaN   
2  Lap Belt & Harness                                                NaN   
3                 NaN                                                NaN   
4                 NaN                                                NaN   

             PED_ACTION                    COMPLAINT    PED_ROLE  \
0  Crossing With Signal  Complaint of Pain or Nausea  Pedestrian   
1                   NaN               Does Not Apply      Driver   
2                   NaN               Does Not Apply   Passenger   
3                   NaN                          NaN  Registrant   
4                   NaN                          NaN         NaN   

   CONTRIBUTING_FACTOR_1 CONTRIBUTING_FACTOR_2 PERSON_SEX  
0            Unspecified           Unspecified          F  
1                    NaN                   NaN          F  
2                    NaN                   NaN          F  
3                    NaN                   NaN          U  
4                    NaN                   NaN        NaN  

[5 rows x 49 columns]

In [7]:
print("The number of records is currently: ",len(df))

The number of records is currently:  4492438


With nearly 4.5 million records, this data at first glance seems large and very expansive, however a good chunk of records do not have spatial data (X,Y) POINT component. These records cannot be used in the analysis since they will not factor into spatial research.

In [8]:
# Removing Values where there is no location data
df1 = df.dropna(subset = ['LOCATION']) # we are subsetting so only records where no location are removed
print("The number of records is currently: ",len(df1))

The number of records is currently:  4012713


In [9]:
# removing seriously erroneous coordinates
df1 = df1[(df1['LATITUDE'] > 38) & (df1['LONGITUDE'] <-48) &
          (df1['LONGITUDE'] > -75)]
        
print("The number of records is currently: ",len(df1))

The number of records is currently:  4007788


Now there are roughly 4.007 million records with a spatial component. The SRID for this data is EPSG:4326. 

In [10]:
pd.options.display.max_columns = None # this is so I can view all 49 columns of data.
df1.head()
# df['CRASH_DATE']= pd.to_datetime(df['CRASH_DATE'])

CRASH DATE CRASH TIME   BOROUGH ZIP CODE  LATITUDE  LONGITUDE  \
0   03/03/2020       0:00  BROOKLYN    11208  40.68992  -73.87261   
12  03/03/2020      13:08  BROOKLYN    11219  40.63693  -73.98657   
13  03/03/2020      13:08  BROOKLYN    11219  40.63693  -73.98657   
14  03/03/2020      13:08  BROOKLYN    11219  40.63693  -73.98657   
15  03/03/2020      13:08  BROOKLYN    11219  40.63693  -73.98657   

                      LOCATION                    ON STREET NAME  \
0   POINT (-73.87261 40.68992)  HEMLOCK STREET                     
12  POINT (-73.98657 40.63693)                               NaN   
13  POINT (-73.98657 40.63693)                               NaN   
14  POINT (-73.98657 40.63693)                               NaN   
15  POINT (-73.98657 40.63693)                               NaN   

   CROSS STREET NAME                           OFF STREET NAME  \
0     JAMAICA AVENUE                                       NaN   
12               NaN  4312      14 AVENUE                        
13               NaN  4312      14 AVENUE                        
14               NaN  4312      14 AVENUE                        
15               NaN  4312      14 AVENUE                        

    NUMBER OF PERSONS INJURED  NUMBER OF PERSONS KILLED  \
0                         1.0                       0.0   
12                        0.0                       0.0   
13                        0.0                       0.0   
14                        0.0                       0.0   
15                        0.0                       0.0   

    NUMBER OF PEDESTRIANS INJURED  NUMBER OF PEDESTRIANS KILLED  \
0                               1                             0   
12                              0                             0   
13                              0                             0   
14                              0                             0   
15                              0                             0   

    NUMBER OF CYCLIST INJURED  NUMBER OF CYCLIST KILLED  \
0                           0                         0   
12                          0                         0   
13                          0                         0   
14                          0                         0   
15                          0                         0   

    NUMBER OF MOTORIST INJURED  NUMBER OF MOTORIST KILLED  \
0                            0                          0   
12                           0                          0   
13                           0                          0   
14                           0                          0   
15                           0                          0   

   CONTRIBUTING FACTOR VEHICLE 1 CONTRIBUTING FACTOR VEHICLE 2  \
0                            NaN                           NaN   
12                   Unspecified                   Unspecified   
13                   Unspecified                   Unspecified   
14                   Unspecified                   Unspecified   
15                   Unspecified                   Unspecified   

   CONTRIBUTING FACTOR VEHICLE 3 CONTRIBUTING FACTOR VEHICLE 4  \
0                            NaN                           NaN   
12                           NaN                           NaN   
13                           NaN                           NaN   
14                           NaN                           NaN   
15                           NaN                           NaN   

   CONTRIBUTING FACTOR VEHICLE 5  COLLISION_ID VEHICLE TYPE CODE 1  \
0                            NaN       4297861                 NaN   
12                           NaN       4297227               Sedan   
13                           NaN       4297227               Sedan   
14                           NaN       4297227               Sedan   
15                           NaN       4297227               Sedan   

   VEHICLE TYPE CODE 2 VEHICLE TYPE CODE 3 VEHICLE TYPE CODE 4  \
0                  NaN

In [12]:
# let's view the data by the highest number of people killed in one incident
sort_person_kill = df1.sort_values(by = 'NUMBER OF PERSONS KILLED', ascending = False)
sort_person_kill.head(10)

CRASH DATE CRASH TIME    BOROUGH ZIP CODE   LATITUDE  LONGITUDE  \
2920724  10/31/2017      15:08  MANHATTAN    10014  40.729046  -74.01073   
2920726  10/31/2017      15:08  MANHATTAN    10014  40.729046  -74.01073   
2920735  10/31/2017      15:08  MANHATTAN    10014  40.729046  -74.01073   
2920736  10/31/2017      15:08  MANHATTAN    10014  40.729046  -74.01073   
2920737  10/31/2017      15:08  MANHATTAN    10014  40.729046  -74.01073   
2920725  10/31/2017      15:08  MANHATTAN    10014  40.729046  -74.01073   
2920723  10/31/2017      15:08  MANHATTAN    10014  40.729046  -74.01073   
2920722  10/31/2017      15:08  MANHATTAN    10014  40.729046  -74.01073   
2920721  10/31/2017      15:08  MANHATTAN    10014  40.729046  -74.01073   
2920720  10/31/2017      15:08  MANHATTAN    10014  40.729046  -74.01073   

                            LOCATION                    ON STREET NAME  \
2920724  POINT (-74.01073 40.729046)  WEST STREET                        
2920726  POINT (-74.01073 40.729046)  WEST STREET                        
2920735  POINT (-74.01073 40.729046)  WEST STREET                        
2920736  POINT (-74.01073 40.729046)  WEST STREET                        
2920737  POINT (-74.01073 40.729046)  WEST STREET                        
2920725  POINT (-74.01073 40.729046)  WEST STREET                        
2920723  POINT (-74.01073 40.729046)  WEST STREET                        
2920722  POINT (-74.01073 40.729046)  WEST STREET                        
2920721  POINT (-74.01073 40.729046)  WEST STREET                        
2920720  POINT (-74.01073 40.729046)  WEST STREET                        

           CROSS STREET NAME OFF STREET NAME  NUMBER OF PERSONS INJURED  \
2920724  WEST HOUSTON STREET             NaN                       12.0   
2920726  WEST HOUSTON STREET             NaN                       12.0   
2920735  WEST HOUSTON STREET             NaN                       12.0   
2920736  WEST HOUSTON STREET             NaN                       12.0   
2920737  WEST HOUSTON STREET             NaN                       12.0   
2920725  WEST HOUSTON STREET             NaN                       12.0   
2920723  WEST HOUSTON STREET             NaN                       12.0   
2920722  WEST HOUSTON STREET             NaN                       12.0   
2920721  WEST HOUSTON STREET             NaN                       12.0   
2920720  WEST HOUSTON STREET             NaN                       12.0   

         NUMBER OF PERSONS KILLED  NUMBER OF PEDESTRIANS INJURED  \
2920724                       8.0                              7   
2920726                       8.0                              7   
2920735                       8.0                              7   
2920736                       8.0                              7   
2920737                       8.0                              7   
2920725                       8.0                              7   
2920723                       8.0                              7   
2920722                       8.0                              7   
2920721                       8.0                              7   
2920720                       8.0                              7   

         NUMBER OF PEDESTRIANS KILLED  NUMBER OF CYCLIST INJURED  \
2920724                             6                          1   
2920726                             6                          1   
2920735                             6                          1   
2920736                             6                          1   
2920737                             6                          1   
2920725                             6                          1   
2920723                             6                          1   
2920722                             6                          1   
2920721                             6                          1   
2920720                             6                          1   

         NUMBER OF CYCLIST KILLED

Using group by(), we can aggregate the collision IDs and treat them each as one event,
adding a count field to dataset to include the total number of people per collision.
TRANSFORM function used to retain the number of rows in our original data-set

In [13]:
df1['count_col'] = df1.groupby(['COLLISION_ID'])[['COLLISION_ID']].transform('count')
df1 = df1.drop_duplicates('COLLISION_ID') # dropping repeat values in COLLISION_ID
df1.head(6)
# [[]] is needed to refer to the data frame column, if it was ['COLLISION_ID'], then it would return a pandas series. 

CRASH DATE CRASH TIME   BOROUGH ZIP CODE   LATITUDE  LONGITUDE  \
0   03/03/2020       0:00  BROOKLYN    11208  40.689920 -73.872610   
12  03/03/2020      13:08  BROOKLYN    11219  40.636930 -73.986570   
29  04/02/2015       8:30       NaN      NaN  40.772880 -73.831198   
39  03/03/2020      17:33    QUEENS    11432  40.708770 -73.790504   
40  10/17/2013      18:30       NaN      NaN  40.640037 -73.877941   
57  09/02/2017      13:25  BROOKLYN    11203  40.650517 -73.923590   

                          LOCATION                    ON STREET NAME  \
0       POINT (-73.87261 40.68992)  HEMLOCK STREET                     
12      POINT (-73.98657 40.63693)                               NaN   
29  POINT (-73.8311985 40.7728803)                               NaN   
39     POINT (-73.790504 40.70877)                               NaN   
40  POINT (-73.8779406 40.6400371)                               NaN   
57     POINT (-73.92359 40.650517)  KINGS HIGHWAY                      

   CROSS STREET NAME                           OFF STREET NAME  \
0     JAMAICA AVENUE                                       NaN   
12               NaN  4312      14 AVENUE                        
29               NaN                                       NaN   
39               NaN  170-12    90 AVENUE                        
40               NaN                                       NaN   
57     SNYDER AVENUE                                       NaN   

    NUMBER OF PERSONS INJURED  NUMBER OF PERSONS KILLED  \
0                         1.0                       0.0   
12                        0.0                       0.0   
29                        0.0                       0.0   
39                        0.0                       0.0   
40                        0.0                       0.0   
57                        1.0                       0.0   

    NUMBER OF PEDESTRIANS INJURED  NUMBER OF PEDESTRIANS KILLED  \
0                               1                             0   
12                              0                             0   
29                              0                             0   
39                              0                             0   
40                              0                             0   
57                              0                             0   

    NUMBER OF CYCLIST INJURED  NUMBER OF CYCLIST KILLED  \
0                           0                         0   
12                          0                         0   
29                          0                         0   
39                          0                         0   
40                          0                         0   
57                          0                         0   

    NUMBER OF MOTORIST INJURED  NUMBER OF MOTORIST KILLED  \
0                            0                          0   
12                           0                          0   
29                           0                          0   
39                           0                          0   
40                           0                          0   
57                           1                          0   

   CONTRIBUTING FACTOR VEHICLE 1 CONTRIBUTING FACTOR VEHICLE 2  \
0                            NaN                           NaN   
12                   Unspecified                   Unspecified   
29       Outside Car Distraction                   Unspecified   
39                   Unspecified                           NaN   
40                   Unspecified                   Unspecified   
57                   Unspecified                           NaN   

   CONTRIBUTING FACTOR VEHICLE 3 CONTRIBUTING FACTOR VEHICLE 4  \
0                            NaN                           NaN   
12                           NaN                           NaN   
29                           NaN                           NaN   
39                           NaN                           NaN   
40                        

The data has been cleaned well, will now be saved as a database file so additional filtering and spatial operations can be done in SQL through PostGIS environment.

In [14]:
# converting zip code to numeric to avoid errors once in PostGIS
df1["ZIP CODE"] = pd.to_numeric(df1["ZIP CODE"], errors = 'coerce')

In [15]:
# after trying to import into PostGIS, these columns were spit back as having mixed types.

# These columns are not necessary for calculations, can be removed.
cols = [3,32,37,38,39,40,41,43,44,45,46,47,48] #specifying columns based on list comprehension
df2 = df1.drop(df1.columns[cols], axis = 1) #dropping the columns


In [16]:
# make our dataframe into a GeoDataframe
df2['LOCATION'] = df2['LOCATION'].apply(wkt.loads) # refer to our LOCATION field as our geometry

gdf = gpd.GeoDataFrame(df2, geometry='LOCATION')
print("The number of records is currently: ",len(gdf))
gdf.head(2)


The number of records is currently:  1458849


CRASH DATE CRASH TIME   BOROUGH  LATITUDE  LONGITUDE  \
0   03/03/2020       0:00  BROOKLYN  40.68992  -73.87261   
12  03/03/2020      13:08  BROOKLYN  40.63693  -73.98657   

                               LOCATION                    ON STREET NAME  \
0   POINT (-73.87260999999999 40.68992)  HEMLOCK STREET                     
12           POINT (-73.98657 40.63693)                               NaN   

   CROSS STREET NAME                           OFF STREET NAME  \
0     JAMAICA AVENUE                                       NaN   
12               NaN  4312      14 AVENUE                        

    NUMBER OF PERSONS INJURED  NUMBER OF PERSONS KILLED  \
0                         1.0                       0.0   
12                        0.0                       0.0   

    NUMBER OF PEDESTRIANS INJURED  NUMBER OF PEDESTRIANS KILLED  \
0                               1                             0   
12                              0                             0   

    NUMBER OF CYCLIST INJURED  NUMBER OF CYCLIST KILLED  \
0                           0                         0   
12                          0                         0   

    NUMBER OF MOTORIST INJURED  NUMBER OF MOTORIST KILLED  \
0                            0                          0   
12                           0                          0   

   CONTRIBUTING FACTOR VEHICLE 1 CONTRIBUTING FACTOR VEHICLE 2  \
0                            NaN                           NaN   
12                   Unspecified                   Unspecified   

   CONTRIBUTING FACTOR VEHICLE 3 CONTRIBUTING FACTOR VEHICLE 4  \
0                            NaN                           NaN   
12                           NaN                           NaN   

   CONTRIBUTING FACTOR VEHICLE 5  COLLISION_ID VEHICLE TYPE CODE 1  \
0                            NaN       4297861                 NaN   
12                           NaN       4297227               Sedan   

   VEHICLE TYPE CODE 2 VEHICLE TYPE CODE 3 VEHICLE TYPE CODE 4  \
0                  NaN                 NaN                 NaN   
12               Sedan                 NaN                 NaN   

   VEHICLE TYPE CODE 5   UNIQUE_ID  CRASH_DATE CRASH_TIME PERSON_TYPE  \
0                  NaN  10583927.0  03/03/2020       0:00  Pedestrian   
12                 NaN  10580837.0  03/03/2020      13:08    Occupant   

   PERSON_INJURY  VEHICLE_ID  PERSON_AGE  \
0        Injured         NaN        21.0   
12   Unspecified  19310666.0        56.0   

                                         PED_LOCATION  count_col  
0   Pedestrian/Bicyclist/Other Pedestrian at Inter...          1  
12                                                NaN          4


PostGIS can be a bit 'finnicky' when importing dataframes. To improve our data,
we will lowercase all columns and remove whitespace, replacing them with '_'

In [17]:
# lowercasing with for loop
gdf.columns = [x.lower() for x in gdf.columns]
# replacing spaces with _
gdf.columns = gdf.columns.str.replace(' ', '_')
gdf.head(5)

crash_date crash_time   borough   latitude  longitude  \
0   03/03/2020       0:00  BROOKLYN  40.689920 -73.872610   
12  03/03/2020      13:08  BROOKLYN  40.636930 -73.986570   
29  04/02/2015       8:30       NaN  40.772880 -73.831198   
39  03/03/2020      17:33    QUEENS  40.708770 -73.790504   
40  10/17/2013      18:30       NaN  40.640037 -73.877941   

                               location                    on_street_name  \
0   POINT (-73.87260999999999 40.68992)  HEMLOCK STREET                     
12           POINT (-73.98657 40.63693)                               NaN   
29       POINT (-73.8311985 40.7728803)                               NaN   
39          POINT (-73.790504 40.70877)                               NaN   
40       POINT (-73.8779406 40.6400371)                               NaN   

   cross_street_name                           off_street_name  \
0     JAMAICA AVENUE                                       NaN   
12               NaN  4312      14 AVENUE                        
29               NaN                                       NaN   
39               NaN  170-12    90 AVENUE                        
40               NaN                                       NaN   

    number_of_persons_injured  number_of_persons_killed  \
0                         1.0                       0.0   
12                        0.0                       0.0   
29                        0.0                       0.0   
39                        0.0                       0.0   
40                        0.0                       0.0   

    number_of_pedestrians_injured  number_of_pedestrians_killed  \
0                               1                             0   
12                              0                             0   
29                              0                             0   
39                              0                             0   
40                              0                             0   

    number_of_cyclist_injured  number_of_cyclist_killed  \
0                           0                         0   
12                          0                         0   
29                          0                         0   
39                          0                         0   
40                          0                         0   

    number_of_motorist_injured  number_of_motorist_killed  \
0                            0                          0   
12                           0                          0   
29                           0                          0   
39                           0                          0   
40                           0                          0   

   contributing_factor_vehicle_1 contributing_factor_vehicle_2  \
0                            NaN                           NaN   
12                   Unspecified                   Unspecified   
29       Outside Car Distraction                   Unspecified   
39                   Unspecified                           NaN   
40                   Unspecified                   Unspecified   

   contributing_factor_vehicle_3 contributing_factor_vehicle_4  \
0                            NaN                           NaN   
12                           NaN                           NaN   
29                           NaN                           NaN   
39                           NaN                           NaN   
40                           NaN                           NaN   

   contributing_factor_vehicle_5  collision_id  \
0                            NaN       4297861   
12                           NaN       4297227   
29                           NaN       3196670   
39                           NaN       4297447   
40                           NaN       2960273   

                    vehicle_type_code_1 vehicle_type_code_2  \
0                                   NaN                 NaN   
12                                Sedan               Sedan   
29           

In [48]:
#save as CSV file
gdf.to_csv(r'filepath\table.csv', index = False)

In [49]:
# importing this data directly into postGIS
hostNameStr = 'localhost'; portNumber  = '5432'
userNameStr = 'postgres' ; passwordStr = 'pw'
databaseStr = 'postgres' ; tableName = 'collision_data'

csvFileName = r'filepath.csv'
# refine path when sharing
#connection format
'dialect+driver://username:password@host:port/database'
dbEngine = create_engine('postgresql://' \
    + userNameStr + ':' + passwordStr \
        + '@' + hostNameStr + ':' + portNumber \
            + '/' + databaseStr)
gdf = pd.read_csv(csvFileName)
gdf.to_sql(tableName, dbEngine)




C:\Users\steve\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3057: DtypeWarning: Columns (35) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [18]:
# check the length of the data
print("The number of records is currently: ",len(gdf))

The number of records is currently:  1458849


There are currently 1,458,849 unique collision records in the data table. From here, we need to aggregate this data based off each individual NYC intersection with collision data.
View part 2 of the project: Spatial Operations in PostGIS